firewood/src/merkle/parallel.rs

```diff
@@ -0,0 +1,500 @@
+}
+
+/// Response returned from a worker to the main thread. Includes the new root of the subtrie
+/// at the given nibble and the deleted nodes.
+#[derive(Debug)]
+struct Response {
+    nibble: u8,
+    root: Option<Child>,
+    deleted_nodes: Vec<MaybePersistedNode>,
+}
+
+#[derive(Debug)]
+pub enum CreateProposalError {
+    FileIoError(FileIoError),
+    SendError,
+}
+
+impl From<FileIoError> for CreateProposalError {
+    fn from(err: FileIoError) -> Self {
+        CreateProposalError::FileIoError(err)
+    }
+}
+
+impl From<SendError<Request>> for CreateProposalError {
+    fn from(_err: SendError<Request>) -> Self {
+        CreateProposalError::SendError
+    }
+}
+
+/// `ParallelMerkle` safely performs parallel modifications to a Merkle trie. It does this
+/// by creating a worker for each subtrie from the root, and allowing the the workers to
+/// perform inserts and removes to their subtries.
+#[derive(Debug)]
+pub struct ParallelMerkle {
+    workers: [Option<WorkerSender>; BranchNode::MAX_CHILDREN],
+}
+
+impl Default for ParallelMerkle {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ParallelMerkle {
+    /// Default constructor
+    #[must_use]
+    pub fn new() -> Self {
+        ParallelMerkle {
+            workers: [(); BranchNode::MAX_CHILDREN].map(|()| None),
+        }
+    }
+
+    /// Normalize the root to allow clean separation of the trie into an array of subtries that
+    /// can be operated on independently by the worker threads.
+    fn normalize_root(
+        &self,
+        proposal: &mut NodeStore<MutableProposal, FileBacked>,
+    ) -> Box<BranchNode> {
+        // There are 3 different cases to handle depending on the value of the root node.
+        //
+        // 1. If root is None, create a branch node with an empty partial path and a None for
+        //    value. Create Nones for all of its children.
+        // 2. If the existing root has a partial path, then create a new root with an empty
+        //    partial path and a None for a value. Push down the previous root as a child. Note
+        //    that this modified Merkle trie is not currently valid and may need to be updated
+        //    during the post-processing step.
+        // 3. If the existing root does not have a partial path, then there is nothing we need
+        //    to do if it is a branch. If it is a leaf, then convert it into a branch.
+        //
+        // Cases 2 and 3 are handled by `normalize_for_insert`. This function returns a branch
+        // node with an empty partial path.
+        let root_node = proposal.root_mut().take();
+        if let Some(node) = root_node {
+            return node.normalize_for_insert();
+        }
+        // Empty trie. Create a branch node with an empty partial path and a None for a value.
+        BranchNode {
+            partial_path: Path::new(),
+            value: None,
+            children: BranchNode::empty_children(),
+        }
+        .into()
```

Comment:
This is an if-let-else or a `map_or_else`

See https://doc.rust-lang.org/book/ch06-03-if-let.html and https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else

```suggestion
        if let Some(node) = root_node {
            node.normalize_for_insert()
        } else {
            // Empty trie. Create a branch node with an empty partial path and a None for a value.
            BranchNode {
                partial_path: Path::new(),
                value: None,
                children: BranchNode::empty_children(),
            }
            .into()
        }
```

or

```
    node.map_or_else(BranchNode{...}.into(), normalize_for_insert)
```